dns/dns.go

```diff
@@ -0,0 +1,124 @@
+package dns
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/giantswarm/exporterkit"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	Namespace = "dns"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+
+	Hosts []string
+}
+
+type DNSCollector struct {
+	logger micrologger.Logger
+
+	hosts []string
+
+	count      map[string]float64
+	errorCount map[string]float64
+
+	countMutex      sync.Mutex
+	errorCountMutex sync.Mutex
+
+	total      *prometheus.Desc
+	errorTotal *prometheus.Desc
+	latency    *prometheus.Desc
+}
+
+func NewCollector(config Config) (*DNSCollector, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Logger must not be empty", config)
```

Comment:
Why is this returning an error of another lib? The problem here is that the constructor of this very package expects a logger. So there should be an error returned of this very package.

Reply:
This is the error design we have. Making up something else now does not benefit our projects. There is an epic in the operator board towards error handling. See https://github.com/giantswarm/giantswarm/issues/3091. Feel free to help improving the situation for all of us. In the meantime I would appreciate complying to the agreements we have so far.